File: DSA/src/Questions/BinarySearch/SingleElement.java
package Questions.BinarySearch;

public class SingleElement {

    public static int singleNonDuplicate(int[] arr) {
        int n = arr.length; // Size of the array.

        int low = 0, high = n - 1;
        while (low <= high) {
            int mid = (high + low) >> 1;

            // If arr[mid] is the single element:
            if ((mid == 0 || arr[mid] != arr[mid - 1]) && (mid == n - 1 || arr[mid] != arr[mid + 1])) {
                return arr[mid];
            }

            // Check if we are in the left part of the pair.
            if (mid % 2 == 0) {
                // If the mid is at an even index, and the next element is the same,
                // then the single element is on the right side.
                if (mid + 1 < n && arr[mid] == arr[mid + 1]) {
                    low = mid + 1;
                } else {
                    // If the mid is at an even index and the next element is different,
                    // then the single element is on the left side.
                    high = mid - 1;
                }
            } else {
                // Check if we are in the right part of the pair.
                // Similar logic as above but in reverse for odd indices.
                if (arr[mid] == arr[mid - 1]) {
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            }
        }

        // Dummy return statement:
        return -1;
    }

    public static void main(String[] args) {
        int[] arr = {1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8 };
        int ans = singleNonDuplicate(arr);
        System.out.println("The single element is: " + ans);
    }
}

File: DSA/Python/Questions/BinarySearch/SingleElement.py
"""
# Question: Single Element in a Sorted Array
# Link: https://leetcode.com/problems/single-element-in-a-sorted-array/

# Time Complexity: O(log n)
# Space Complexity: O(1)

# Algorithm:
# 1. Binary search with pair checking
# 2. Find pattern break
# 3. Return single element
"""


class SingleElement:
    def single_non_duplicate(self, nums: list[int]) -> int:
        left, right = 0, len(nums) - 1

        while left < right:
            mid = (left + right) >> 1

            # Ensure mid is even
            if mid % 2 == 1:
                mid -= 1

            if nums[mid] == nums[mid + 1]:
                left = mid + 2
            else:
                right = mid

        return nums[left]


def main():
    solution = SingleElement()
    nums = [1, 1, 2, 3, 3, 4, 4, 8, 8]
    print(solution.single_non_duplicate(nums))  # Expected: 2


if __name__ == "__main__":
    main()
